vector/v.info/print.c

```diff
@@ -198,30 +199,92 @@ void print_columns(struct Map_info *Map, const char *input_opt,
                 "layer <%s>:"),
               field_opt);
 
-    if ((fi = Vect_get_field2(Map, field_opt)) == NULL)
+    if ((fi = Vect_get_field2(Map, field_opt)) == NULL) {
+        Vect_close(Map);
         G_fatal_error(
             _("Database connection not defined for layer <%s> of <%s>"),
             field_opt, input_opt);
+    }
     driver = db_start_driver(fi->driver);
-    if (driver == NULL)
+    if (driver == NULL) {
+        Vect_close(Map);
         G_fatal_error(_("Unable to open driver <%s>"), fi->driver);
+    }
     db_init_handle(&handle);
     db_set_handle(&handle, fi->database, NULL);
-    if (db_open_database(driver, &handle) != DB_OK)
+    if (db_open_database(driver, &handle) != DB_OK) {
+        db_shutdown_driver(driver);
+        Vect_close(Map);
         G_fatal_error(_("Unable to open database <%s> by driver <%s>"),
                       fi->database, fi->driver);
+    }
     db_init_string(&table_name);
     db_set_string(&table_name, fi->table);
-    if (db_describe_table(driver, &table_name, &table) != DB_OK)
+    if (db_describe_table(driver, &table_name, &table) != DB_OK) {
+        db_close_database(driver);
+        db_shutdown_driver(driver);
+        Vect_close(Map);
         G_fatal_error(_("Unable to describe table <%s>"), fi->table);
+    }
+
+    JSON_Value *root_value = NULL, *columns_value = NULL;
+    JSON_Object *root_object = NULL;
+    JSON_Array *columns_array = NULL;
+
+    if (format == JSON) {
+        root_value = json_value_init_object();
+        root_object = json_object(root_value);
+        columns_value = json_value_init_array();
+        columns_array = json_array(columns_value);
+        json_object_set_value(root_object, "columns", columns_value);
+    }
 
     ncols = db_get_table_number_of_columns(table);
-    for (col = 0; col < ncols; col++)
-        fprintf(stdout, "%s|%s\n",
-                db_sqltype_name(
-                    db_get_column_sqltype(db_get_table_column(table, col))),
+    for (col = 0; col < ncols; col++) {
+        if (format == JSON) {
```

Comment:
Please consider using `switch` here and in similar cases. Especially, when we have an enum already in `format`. It leads in my opinion to clearer code and have the advantage in that compilers may detect flaws (eg. -Wswitch-enum).